portal-api/src/v1/controllers/countries.controller.js

```diff
@@ -1,44 +1,122 @@
 const utils = require('../../utils/array');
-
 const externalApi = require('../../external-api/api');
 
-const getCountryFromArray = (arr, code) => arr.filter((country) => country.code === code)[0];
+/**
+ * Retrieves a country object from an array based on the provided country code.
+ *
+ * @param {Array} arr - The array of country objects to search in.
+ * @param {string} code - The country code to search for.
+ * @returns {Array} - The country object matching the provided code, or an empty array if not found.
+ */
+const getCountryFromArray = (arr, code) => (arr.length ? arr.filter((country) => country.code === code)[0] : []);
 
+/**
+ * Sorts an array of countries.
+ *
+ * @param {Array} countries - The array of countries to be sorted.
+ * @returns {Array} - The sorted array of countries.
+ * @throws {Error} - If unable to sort countries.
+ */
 const sortCountries = (countries) => {
-  const countriesWithoutUK = countries.filter((country) => country.code !== 'GBR');
+  try {
+    const countriesWithoutUK = countries.filter((country) => country.code !== 'GBR');
 
-  const sortedArray = [
-    getCountryFromArray(countries, 'GBR'),
-    ...utils.sortArrayAlphabetically(countriesWithoutUK, 'name'),
-  ];
+    const sortedArray = [getCountryFromArray(countries, 'GBR'), ...utils.sortArrayAlphabetically(countriesWithoutUK, 'name')];
 
-  return sortedArray;
+    return sortedArray;
+  } catch (error) {
+    console.error('Unable to sort countries %o', error);
+    throw new Error('Unable to sort countries', { cause: error });
```

Comment:
Have refactored to return an error rather than throwing one.

Reply:
Sorry -- by this I mean we should still be throwing an error (`throw new bla`), but we should look to throw a specific error ie `throw new UnableToSortCountriesError`. This would allow us to catch these specific errors in the controller function ie:


in SortCountriesError.js
```
class SortCountriesError extends Error {
  constructor({cause = undefined }) {
    const message = `Failed to sort countries`;
    super(message, { cause });
    this.name = this.constructor.name;
  }
}
```

in sortCountries function
```
  } catch (error) {
    // we dont need to log anything here as we'll do it in the controller level, shown later
    throw new SortCountriesError({ cause: error });
```

in getCountry funciton

```
const getCountry = async (code) => {
    const response = await externalApi.countries.getCountry(code);

    if (!response) {
      throw new CountryNotFoundError(code);
    }

    return response;
    // we dont need to catch anything here as we'll do it in the controller level
};
```

in controller function -- 
```
      // in the catch block
      console.error('Unable to findAll (or whatever it is) %o', error); // This will return the cause of the error in the object too -- so we know which endpoint got called to cause the error and where the error occured
      if (e instanceof SortCountriesError) {
        // we can then handle this however we want, but do it
        // specifically for this type of error.
        // in this case it will probably return a 500.
      }

      if (e instanceof CountryNotFoundError) {
        // we can then handle this error seperately
        // in this case it is probably a 400 if it returns a void
      }
```

The existing implimentation means that we have to know to check length of the response -- and we're relying on error of `sortCountries` that `{error:"foo"}` returns `undefined` when checking the length of the response. This would mean that sort countries erroring would return a `400` (on loc 116) and get countries returning a `500`. I'll drop a phone about this comment as I think it's easier to talk about.

Reply:
As discussed will handle this in a separate PR.